### PR TITLE
add optional defmt support for error types

### LINKED ISSUE #75: Feature Request: Support defmt Formatting
Working on a an embedded project where there is the ability to log errors and structures with defmt, this does require adding a new conditional dependency and adding a conditional derive as follows. This would be particularly useful for reporting errors from JSON de-serialization.
```rust
#[cfg_attr(feature = "defmt", derive(defmt::Format))]
pub enum Error {
...
}
```

### PR DESCRIPTION
**Pull request by avlec**
add defmt as an optional dependent package

I am only able to test defmt v0.3.0 as the rest of my project is using that version the latest is at v0.3.5

closes #75

### CODE REVIEW

**Review by ryan-summers — COMMENTED** [review]
Changes look good me to, thanks for adding it! Can you add an entry in `CHANGELOG.md` mentioning the change?

**Review by ryan-summers — COMMENTED** [review]
Also, it would be good to run CI with `--all-features`, so maybe update `.github/workflows/ci.yml` as well if you're comfortable. If not, let me know and I'll do it :)

**Comment by avlec** [comment]
I hope I did the CI part properly if not feel free to amend that...

File: src/de/mod.rs [hunk 1]
```diff
@@ -19,6 +19,7 @@ pub type Result<T> = core::result::Result<T, Error>;
 /// This type represents all possible errors that can occur when deserializing JSON data
 #[derive(Debug, PartialEq, Eq, Clone)]
 #[cfg_attr(not(feature = "custom-error-messages"), derive(Copy))]
+#[cfg_attr(feature = "defmt", derive(defmt::Format))]
```
Comment (ryan-summers) [inline src/de/mod.rs:22]: I think you need to add:
```rs
        #[defmt(Debug2Format)]
```

above the `heapless::String<64>` error type.

Comment (Georges760) [reply src/de/mod.rs:22]: or adding "defmt-impl" into the heapless dependency

**Comment by vhdirk** [comment]
Is there anything here I can help with to get this merged?

**Comment by ryan-summers** [comment]
@vhdirk At this point, I would consider this PR abandoned. If you want to fork this current work and address my remaining comments, I'd be happy to review it as a separate PR :)

**Comment by Georges760** [comment]
@vhdirk if you can't do it, i can also help, tell me

**Comment by ryan-summers** [comment]
It's already being continued in https://github.com/rust-embedded-community/serde-json-core/pull/77 for now :)

**Comment by avlec** [comment]
@vhdirk Thanks for picking up where I started with this and getting it in there!

**Pull request closed without merging**